### PR TITLE
dbGetChunkedQuery

### PR DESCRIPTION
**Pull request by bborgesr**
Moving the relevant code for the `dbGetChunkedQuery()` function to this PR, originally in rstats-db/DBI#108

### CODE REVIEW

File: R/DBConnection.R [hunk 1]
```diff
@@ -106,6 +106,97 @@ setGeneric("dbSendQuery",
   valueClass = "DBIResult"
 )
 
+#' Sentinel value to indicate the early termination of
+#' \code{\link{dbGetChunkedQuery}}
+#' @export
+dbBreak <- structure(list(), class = "dbi_break_sentinel")
```
Comment (hadley) [inline R/DBConnection.R]: Needs to be documented with `dbGetChunkedQuery()`

File: R/DBConnection.R [hunk 2]
```diff
@@ -106,6 +106,97 @@ setGeneric("dbSendQuery",
...
+
+dbFetchChunked <- function(rs, n, callback) {
+  while (TRUE) {
+    chunk <- dbFetch(rs, n = n)
+    if (nrow(chunk) == 0) {
+      return(chunk)
+    }
+    result <- callback(chunk)
```
Comment (hadley) [inline R/DBConnection.R]: I wonder if it would be useful to include the cumulative number of rows in the callback? Or the number of times it's been called?

And I _think_, the call should be `callback(chunk = chunk)` to force people to use a consistently named argument. That  makes it easier to add new arguments in the future. But @jcheng5 will have better advice based on shiny experience.

File: R/DBConnection.R [hunk 3]
```diff
@@ -106,6 +106,97 @@ setGeneric("dbSendQuery",
...
+#' dbGetChunkedQuery(con, "SELECT * FROM cars", 1, function(df) {
+#'   rowCount <<- rowCount + nrow(df)
+#'   if (df$speed == 19 && df$dist == 46) {
+#'     print(paste("Your row is number", rowCount))
+#'     return(dbBreak)
+#'   }
+#' })
+#'
+#' dbDisconnect(con)
+setGeneric("dbGetChunkedQuery", function(conn, statement, n, callback, ...) {
```
Comment (hadley) [inline R/DBConnection.R]: I think maybe that `callback` should come before `n`, so `n` can get a reasonable default value, say 1e4

Comment (hadley) [inline R/DBConnection.R]: I think you should probably also specify that this function is generic only on `conn` and `statement` by specifying the `signature` argument to `setGeneric`. I _think_ that's best practice.

File: R/DBConnection.R (hunk 2 above)
Comment (bborgesr) [inline R/DBConnection.R]: @hadley, I'm not sure what you mean by "include the cumulative number of rows in the callback"? As in, when the fetching is over (or alternatively, when the user interrupts the function), print a message indicating how many rows were used (or how many times the callback was called)?

Comment (jcheng5) [inline R/DBConnection.R]: I think Hadley meant `callback(chunk, index = rowsSoFar)` so that the callback can know how far it is into the iteration.

Comment (jcheng5) [inline R/DBConnection.R]: Personally, I prefer `callback(chunk)` to `callback(chunk = chunk)` because I like giving people flexibility to call the "main" parameter whatever they like--in some cases their code might be more readable if they can say `function(transactions) {` instead of `function(chunk) {`.. However I do think it's a good idea to call with the name when talking about other (minor) arguments.

File: R/DBConnection.R (hunk 3 above)
Comment (jcheng5) [inline R/DBConnection.R]: What can `statement` be besides a string? (Future prepared-statement object?)

Comment (bborgesr) [inline R/DBConnection.R]: @hadley, I didn't realize that was necessary (specifying `signature` on `setGeneric`), but it does make sense so I changed it. As for the order of the params: if you look at the example the function is specified as an anonymous function, directly on the params' vector. I figured that would be the most common use, and, in that case, we wouldn't want to have other arguments after it. Otherwise, I agree, it would make sense for `n` to be the last one. Should I change the order anyway?

Comment (jcheng5) [inline R/DBConnection.R]: Even with the callback an anonymous function, I think that `n = 10000L` should come after.

Comment (hadley) [inline R/DBConnection.R]: @jcheng5 yeah, could be a prepared query object, or you could imagine it being a connection (to SQL in a file) or .. ?  Probably not super important, but I think it's dispatchable in other DBI generics.

File: R/DBConnection.R [hunk 4]
```diff
@@ -106,6 +106,39 @@ setGeneric("dbSendQuery",
   valueClass = "DBIResult"
 )
 
+#' @export
+dbBreak <- structure(list(), class = "dbi_break_sentinel")
+
+dbFetchChunked <- function(rs, n, callback) {
```
Comment (krlmlr) [inline R/DBConnection.R]: Something like dbProcessChunked() looks like a more suitable name to me. I thought dbFetchChunked() is supposed to fetch incrementally but return a single big data frame consisting of all chunks. (Perhaps dbFetchChunked() can be implemented using dbProcessChunked()?)

File: R/DBConnection.R [hunk 5]
```diff
@@ -106,6 +106,39 @@ setGeneric("dbSendQuery",
...
+  while (TRUE) {
```
Comment (krlmlr) [inline R/DBConnection.R]: Do we want to specify a total maximum number of rows we want to process?

File: R/DBConnection.R [hunk 6]
```diff
@@ -106,6 +106,39 @@ setGeneric("dbSendQuery",
...
+    chunk <- dbFetch(rs, n = n)
+    if (nrow(chunk) == 0) {
+      return(chunk)
+    }
+    result <- callback(df = chunk)
+    if (identical(result, dbBreak)) {
+      return(chunk)
+    }
+  }
```
Comment (krlmlr) [inline R/DBConnection.R]: What's a sensible return value here? Perhaps the results of all callbacks could be accumulated in a list, processed by a second callback, and then returned?

```
dbProcessChunked(..., callback, process_results, ...)
  while (...) {
    ...
    all_results <- c(all_results, list(callback(...)))
    ...
  }
  process_results(all_results)
}

File: R/DBConnection.R [hunk 7]
```diff
@@ -106,6 +106,39 @@ setGeneric("dbSendQuery",
...
+dbBreak <- structure(list(), class = "dbi_break_sentinel")
+
+dbFetchChunked <- function(rs, n, callback) {
+  while (TRUE) {
+    chunk <- dbFetch(rs, n = n)
+    if (nrow(chunk) == 0) {
+      return(chunk)
+    }
+    result <- callback(df = chunk)
+    if (identical(result, dbBreak)) {
```
Comment (krlmlr) [inline R/DBConnection.R]: How about catching conditions of a specific class, instead of relying on a sentinel?

```
  continue_loop <- tryCatch({ result <- callback(...); TRUE }, dbi_abort = function(e) FALSE)
```

File: R/DBConnection.R (hunk 6 above)
Comment (hadley) [inline R/DBConnection.R]: That seems overly complicated to me - I think this function should be called primarily for its side effects.

File: R/DBConnection.R (hunk 7 above)
Comment (jcheng5) [inline R/DBConnection.R]: Yeah you're right, that seems better.

File: R/DBConnection.R (hunk 4 above)
Comment (jcheng5) [inline R/DBConnection.R]: (BTW I think `dbFetchChunked` was supposed to be named `dbFetchChunkedQuery` and exported.)

I see what you mean, but I like the `dbGetQuery => dbGetChunkedQuery`, `dbFetchQuery => dbFetchChunkedQuery` parallel. Think of the "chunked" modifier as saying instead of getting a result back, you receive the results incrementally through a callback. (Maybe "chunked" should be suffix instead of in the middle, i.e. `dbGetQueryChunked`.) This "chunked" stuff is a general pattern we should make greater use of in the R community, IMHO. `read_csv`, `httr::GET`, for examples.

In other words, don't think of it as changing from "fetching" to "processing"--you're still fetching, just receiving the results in a  different way.

Comment (hadley) [inline R/DBConnection.R]: For a bit more context, the goal is that once we get the API right here, it'll gradually get added to readr, readxl, haven etc.

File: R/DBConnection.R (hunk 2 above)
Comment (bborgesr) [inline R/DBConnection.R]: I'm adding an extra argument to the callback to allow for `callback(chunk, index = rowsSoFar)`. But this got me thinking, are we okay with the fact that the callback cannot take any extra arguments at all (no `...`)? The way things are currently implemented, this is not possible because the `...` are taken up by the `dbSendQuery` method:

```r
#' @export
setMethod("dbGetChunkedQuery", signature("DBIConnection", "character"),
  function(conn, statement, callback, n, ...) {
    rs <- dbSendQuery(conn, statement, ...)
    on.exit(dbClearResult(rs))
    dbFetchChunked(rs, callback, n)
  }
)
```

File: R/DBConnection.R (hunk 4 above)
Comment (bborgesr) [inline R/DBConnection.R]: @jcheng5 I get the name change, but I'm less sure on the need for `dbFetchChunkedQuery` to be exported. Unlike `dbFetch`, this is *not* something that individual backends should implement themselves. This is simply a pattern of fetching that uses the `dbFetch`, which is of course implemented by each backend. So no need for export there (it might actually confuse backend authors in my opinion). This is also not something that the end user should be calling directly -- that is why `dbGetChunkedQuery` exists (which is essentially just a wrapper around `dbFetchChunkedQuery`, which guarantees the closing of the result set). So no need for exporting here either. Am I missing something?

**Comment by bborgesr** [comment]
Barring any outstanding issues that might still continue in this conversation, I think this looks more or less done for DBI. From what I understand, @jcheng5 and @hadley want to generalize this pattern some more before we push it onto master, correct? Let me know need me to do anything further here. And thanks everyone for the great feedback! This was a tough one to get right...

**Comment by krlmlr** [comment]
Could you please merge/rebase with master?

**Comment by bborgesr** [comment]
@krlmlr, rebased :)

File: R/DBConnection.R [hunk 8]
```diff
@@ -106,6 +106,41 @@ setGeneric("dbSendQuery",
   valueClass = "DBIResult"
 )
 
+#' @export
+dbBreak <- structure(list(), class = "dbiAbort")
```
Comment (krlmlr) [inline R/DBConnection.R]: This probably should be a "proper" condition object with `class = c("dbiAbort", "error", "condition")`, see `dput(simpleError("message"))` . Also, I guess dbBreak() should be a function that raises such a condition.

File: man/dbGetChunkedQuery.Rd [hunk 9]
```diff
@@ -0,0 +1,87 @@
+\docType{data}
```
Comment (krlmlr) [inline man/dbGetChunkedQuery.Rd]: I prefer this file to be generated by roxygen2. Rd2roxygen can help here.

File: R/DBConnection.R (hunk 8 above)
Comment (bborgesr) [inline R/DBConnection.R]: I'm not sure that was the intention behind this... This seems like a discussion worth having. The idea was for `dbBreak` to be analogous to `break`: i.e. a basic-looking control statement that allows you to get out of the `while` loop inside `dbFetchChunkedQuery`. It would be great, of course, to be able to simply use `break` here, but that doesn't work. It still sucks that you have to actually return `dbBreak`, rather than just invoke it; but oftentimes that will turn out to look exactly the same. Even when it doesn't, that construct still feels more intuitive to me from a user perspective than having to call a function to break out of the loop... So, even though we use `tryCatch` logic to catch this, it really shouldn't be thought of as a condition object, but as a control statement (or, more accurately, a sentinel value that mimics a control statement)...

That being said, I understand that this may just be confusing. @jcheng5, @hadley , thoughts?

File: man/dbGetChunkedQuery.Rd (hunk 9 above)
Comment (bborgesr) [inline man/dbGetChunkedQuery.Rd]: done

File: R/DBConnection.R (hunk 8 above)
Comment (krlmlr) [inline R/DBConnection.R]: With your implementation of `dbBreak`, you'd have to say `stop(dbBreak)` to get out of the loop. I think `dbBreak()` is better for hiding the implementation details, which should not be the user's concern. We also may allow a custom error message in `dbBreak()`.

File: R/DBConnection.R [hunk 10]
```diff
@@ -106,6 +106,108 @@ setGeneric("dbSendQuery",
...
+
+dbFetchChunkedQuery <- function(rs, callback, n) {
+  rowsSoFar <- 0
+  continueLoop <- TRUE
```
Comment (krlmlr) [inline R/DBConnection.R]: I'm not sure we need `continueLoop` here at all.

File: R/DBConnection.R (hunk 8 above)
Comment (bborgesr) [inline R/DBConnection.R]: Ah, I think I understand better your objection now. You're right that the particular way I had designed the function wasn't working quite right. (I hadn't realized that `tryCatch` would only be triggered in response to an existing condition -- i.e. `stop()`, `warning()` or `message()`.) So I updated it (and actually simplified it!) to work how it's supposed to, which is analogous to a regular `break` statement. The implementation details are equally hidden, I think. Instead of telling the user to call a specific function to break out the the function, we tell them to return a specific value.

So, again, I think we're at the same impasse as before -- do we want this to be analogous to a control statement or, rather, a regular function that raises a condition?

As for the custom error message, I think it's the same thing: if `dbBreak` functions as `break` does, it doesn't make sense for it to be associated with an error message. In fact, in that case, it shouldn't be seen as an error at all -- rather, it's the user expectedly controlling the execution of the code using a control statement, rather than unexpectedly interrupting the execution of the code because of an error.

Since this conception of how `dbBreak` was to work was the result of a discussion with @jcheng5 (and I think @hadley, though I can't quite recall), I'd prefer to get their input before we change this. However, if you feel strongly about this and no-one else pitches in, I'm okay changing things over. I personally think this is a more intuitive and idiomatic way to code a function like this, but I'm not strongly opposed to the alternative.

Comment (jcheng5) [inline R/DBConnection.R]: Ahh shoot. @jimhester, @hadley and I have actually made a lot of progress with chunked API design that we probably need to brain dump before this PR can proceed.

I think in our original discussion we preferred return(dbBreak()) or stop(dbBreak()) because it made it clearer to the reader that normal control flow would be interrupted. But now that I think about it, it'll probably be fairly obvious in most cases anyway. 

```r
callback = function(df, i) {
  if (someTest(df)) {
    dbBreak()
  }
  # more processing
}
```

So I'm ok either way.

Comment (krlmlr) [inline R/DBConnection.R]: `dbWithTransaction()` is another place where we may want to be able to break out cleanly (and roll back) (#115). I'd like to implement only one way of breaking out for both chunked queries and transactions, and whatever may come. In `dbFetchChunkedQuery()` there is an isolated function where the user can `return()`, but in `dbWithTransaction()` there's just a code block, and using a sentinel value seems difficult in many cases -- the condition system seems to be suited well for this situation.

Is everybody okay with

```r
dbBreak <- function() {
  signalCondition(
    structure(
      list(message = "Aborting DBI processing", call = NULL),
      class = c("dbiAbort", "condition")))
  stop("Invalid usage of dbBreak().", call. = FALSE)
}
```

and corresponding code in `dbFetchChunkedQuery()`?

Comment (jcheng5) [inline R/DBConnection.R]: It's fine with me. I think I may need to sync up with @bborgesr as we discussed it on slack and are not totally on the same page yet.

Comment (bborgesr) [inline R/DBConnection.R]: I've looked over `readr` and love it. I think we should definitely adopt that API here, I'm working on a rough adaptation for us to go over. *But* if we do this, I'm not sure that the breaking from the `chunked` stuff should be same construct as the breaking from other DBI stuff.

In particular, the pattern used in `readr` is simply: if the callback returns `FALSE`, stop the iteration. What do you think @krlmlr, @jcheng5, should we go with this or something more explicit?

File: R/callback.R [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+#' rowCount <- 0
+#' callback <- function(df, index) {
+#'   distSum <<- distSum + sum(df$dist)
+#'   rowCount <<- rowCount + nrow(df)
+#' }
+#' rs <- dbGetChunkedQuery(con, "SELECT * FROM cars",
+#'   DataFrameCallback$new(callback), n = 10)
+#'
+#' dbDisconnect(con)
+ChunkCallback <- R6::R6Class("ChunkCallback",
```
Comment (krlmlr) [inline R/callback.R]: This requires importing R6, I guess @hannesmuehleisen will hate it.

Comment (bborgesr) [inline R/callback.R]: Eventually this will be a package called `chunked` or the like, but it will certainly use R6 (at least, that was my understanding), so either directly or indirectly DBI will have to import R6 if it wants to be part of this standardized chunked API...

Comment (krlmlr) [inline R/callback.R]: I'm fine to have this functionality in a separate package which imports DBI, but I'd prefer to have DBI free of import dependencies. (The backends themselves are free to import whatever they like, though.)

**Comment by curtisalexander** [comment]
Sorry to drop this comment within a PR...but is there a central discussion that I can follow where discussion centers on a `chunked` package?  Or is the conversation taking place within each project / package: `readxl`, `readr`, `haven`, `DBI`, etc?  

I'm very interested in a `chunked` and/or streaming package that has a unified API across the tidyverse.

**Comment by hadley** [comment]
What's the problem with adding a dependency on R6?  We could make it only suggested, since it's not absolutely required to use DBI.  I think it's better to be consistent across all the packages that will support a chunked interface (and we will work on a uniform chunked package)

**Comment by krlmlr** [comment]
I don't mind a suggests-only dependency.

**Comment by krlmlr** [comment]
Closing for now.
